File: 125.Valid-Palindrome.py
# https://leetcode.com/problems/valid-palindrome/
#
# algorithms
# Easy (30.75%)
# Total Accepted:    341,766
# Total Submissions: 1,111,478


class Solution(object):
    def isPalindrome(self, s):
        """
        :type s: str
        :rtype: bool
        """
        s = s.lower()
        length = len(s)

        l, r = 0, length - 1
        while l < r:
            while l < r and not self.is_valid(s[l]):
                l += 1
            while l < r and not self.is_valid(s[r]):
                r -= 1
            if l == r:
                return True

            if s[l] != s[r]:
                return False
            l += 1
            r -= 1

        return True

    def is_valid(self, ch):
        ascii = ord(ch)
        if 48 <= ascii <= 57 or 97 <= ascii <= 122:
            return True

        return False

File: 125.Valid-Palindrome.java
// https://leetcode.com/problems/valid-palindrome/
//
// algorithms
// Easy (33.42%)
// Total Accepted: 457,696
// Total Submissions: 1,369,350


class Solution {

    public boolean isPalindrome(String s) {
        int len = s.length();
        char[] chs = s.toLowerCase().toCharArray();

        int i = 0, j = len - 1;
        while (i < j) {
            while (i < j && !isLetterOrNum(chs[i])) {
                i++;
            }

            while (i < j && !isLetterOrNum(chs[j])) {
                j--;
            }

            if (i < j && chs[i] != chs[j]) {
                return false;
            }

            i++;
            j--;
        }

        return true;
    }

    public boolean isLetterOrNum(char ch) {
        if (ch >= 'a' && ch <= 'z') {
            return true;
        }

        if (ch >= '0' && ch <= '9') {
            return true;
        }

        return false;
    }

}